### PR TITLE
Np 47019 SearchNviCandidatesHandler, add ViewingScopeValidator

### PR DESCRIPTION
**Pull request by monaullah**
Jira task: NP-47019 Ensure curator only see NVI candidates for their viewingScope

Changes
- In stead of just asserting user is in same institution as requested affiliations, also assert requested affiliations is in users viewingScope
- Added `ViewingScopeValidator`
- Added `UserRetriever` to fetch `User` with `viewingScope` from identity service

One Codacy issue that I'm letting stay: _"Linguistics Antipattern - The field 'hasPart' indicates linguistically it is a boolean, but it is 'List'"_ on `Organization.java`. Let me know if I should fix this.

### CODE REVIEW

File: nvi-commons/src/main/java/no/sikt/nva/nvi/common/validator/ViewingScopeValidator.java [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+    private final OrganizationRetriever organizationRetriever;
+
+    public ViewingScopeValidator(UserRetriever userRetriever,
+                                 OrganizationRetriever organizationRetriever) {
+
+        this.userRetriever = userRetriever;
+        this.organizationRetriever = organizationRetriever;
+    }
+
+    public boolean userIsAllowedToAccess(String userName, List<URI> requestedOrganizations) {
```
Comment (monaullah) [inline nvi-commons/src/main/java/no/sikt/nva/nvi/common/validator/ViewingScopeValidator.java:31]: Most important method in PR

File: nvi-commons/src/main/java/no/sikt/nva/nvi/common/client/model/User.java [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+    public static User fromJson(String jsonString) {
+        return attempt(() -> dtoObjectMapper.readValue(jsonString, User.class)).orElseThrow();
+    }
+
+    public record ViewingScope(@JsonProperty("includedUnits") List<String> includedUnits) {
+
+        public Set<URI> getIncludedUnitUris() {
+            return includedUnits.stream().map(URI::create).collect(Collectors.toSet());
+        }
+    }
```
Comment (sondrev) [inline nvi-commons/src/main/java/no/sikt/nva/nvi/common/client/model/User.java:23]: I belive you also need to account for excluded units

**Review by sondrev — CHANGES_REQUESTED** [review]

**Review by sondrev — COMMENTED** [review]
Ignore previous request-changes

File: nvi-commons/src/main/java/no/sikt/nva/nvi/common/client/UserRetriever.java [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+                   .orElseThrow();
+    }
+
+    private URI createUri(String userName) {
+        return UriWrapper.fromUri(apiHost)
+                   .addChild(USERS_ROLES_PATH, USERS_PATH)
+                   .addChild(userName)
+                   .getUri();
+    }
+}
```
Comment (sondrev) [inline nvi-commons/src/main/java/no/sikt/nva/nvi/common/client/UserRetriever.java:56]: in commons there is a IdentityServiceClient. It does not have a fetchUsers method, but you could add it :)

File: nvi-commons/src/main/java/no/sikt/nva/nvi/common/validator/ViewingScopeValidator.java [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+        var viewingScope = fetchViewingScope(userName);
+        var allowed = getAllowedUnits(viewingScope);
+        var illegal = difference(allowed, requestedOrganizations);
```
Comment (sondrev) [inline nvi-commons/src/main/java/no/sikt/nva/nvi/common/validator/ViewingScopeValidator.java:34]: Sets.difference(), if you dont need a new dependency

File: nvi-commons/src/main/java/no/sikt/nva/nvi/common/client/UserRetriever.java (hunk 3 above)
Comment (monaullah) [reply nvi-commons/src/main/java/no/sikt/nva/nvi/common/client/UserRetriever.java:56]: I didn't know! Thanks for tips, will check out!

Comment (monaullah) [reply nvi-commons/src/main/java/no/sikt/nva/nvi/common/client/UserRetriever.java:56]: https://github.com/BIBSYSDEV/nva-commons/pull/500

**Review by brinxmat — APPROVED** [review]

**Pull request merged**